File: src/main/java/me/melijn/jda/commands/util/StatsCommand.java
package me.melijn.jda.commands.util;

import com.sun.management.OperatingSystemMXBean;
import me.melijn.jda.blub.Category;
import me.melijn.jda.blub.Command;
import me.melijn.jda.blub.CommandEvent;
import me.melijn.jda.utils.Embedder;
import net.dv8tion.jda.bot.sharding.ShardManager;

import java.lang.management.ManagementFactory;
import java.text.DecimalFormat;
import java.util.concurrent.ThreadPoolExecutor;

import static me.melijn.jda.Melijn.PREFIX;

public class StatsCommand extends Command {

    public StatsCommand() {
        this.commandName = "stats";
        this.description = "Shows the bot's server statistics";
        this.usage = PREFIX + commandName;
        this.category = Category.UTILS;
        this.id = 94;
    }

    /* CREDITS TO DUNCTE123 FOR SOME OF THESE STATS AND DESIGN */

    @Override
    protected void execute(CommandEvent event) {
        OperatingSystemMXBean bean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean.class);
        long totalMem = bean.getTotalPhysicalMemorySize() >> 20;
        long usedMem = totalMem - (bean.getFreePhysicalMemorySize() >> 20);
        long totalJVMMem = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getMax() >> 20;
        long usedJVMMem = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage().getUsed() >> 20;
        long voiceChannels = event.getJDA().asBot().getShardManager().getShards().stream().mapToLong(
                (shard) -> shard.getVoiceChannels().stream().filter(
                        (vc) -> vc.getMembers().contains(vc.getGuild().getSelfMember())
                ).count()
        ).sum();
        ThreadPoolExecutor threadPoolExecutor = (ThreadPoolExecutor) event.getClient().getMelijn().getTaskManager().getExecutorService();
        ThreadPoolExecutor scheduledExecutorService = (ThreadPoolExecutor) event.getClient().getMelijn().getTaskManager().getScheduledExecutorService();

        ShardManager shardManager = event.getJDA().asBot().getShardManager();
        event.reply(new Embedder(event.getVariables(), event.getGuild())
                .setThumbnail(event.getJDA().getSelfUser().getAvatarUrl())
                .addField("Bot stats", "" +
                        "\n**Shards** " + shardManager.getShardsTotal() +
                        "\n**Unique users** " + shardManager.getUserCache().size() +
                        "\n**Guilds** " + shardManager.getGuildCache().size() +
                        "\n**Connected VoiceChannels** " + voiceChannels +
                        "\n**Threads** " + (threadPoolExecutor.getActiveCount() + scheduledExecutorService.getActiveCount() + scheduledExecutorService.getQueue().size()) +
                        "\n**Uptime** " + event.getMessageHelper().getDurationBreakdown(ManagementFactory.getRuntimeMXBean().getUptime()) +
                        "\n\u200B", false)
                .addField("Server Stats", "" +
                        "\n**Cores** " + bean.getAvailableProcessors() +
                        "\n**RAM Usage** " + usedMem + "MB/" + totalMem + "MB" +
                        "\n**System Uptime** " + event.getMessageHelper().getDurationBreakdown(event.getHelpers().getSystemUptime()) +
                        "\n\u200B", false)
                .addField("JVM Stats", "" +
                        "\n**CPU Usage** " + new DecimalFormat("###.###%").format(bean.getProcessCpuLoad()) +
                        "\n**RAM Usage** " + usedJVMMem + "MB/" + totalJVMMem + "MB" +
                        "\n**Threads** " + Thread.activeCount() + "/" + Thread.getAllStackTraces().size(), false)
                .build());
    }
}

File: src/main/java/me/melijn/melijnbot/commands/utility/StatsCommand.kt
package me.melijn.melijnbot.commands.utility

import com.sun.management.OperatingSystemMXBean
import me.melijn.melijnbot.objects.command.AbstractCommand
import me.melijn.melijnbot.objects.command.CommandCategory
import me.melijn.melijnbot.objects.command.CommandContext
import me.melijn.melijnbot.objects.embed.Embedder
import me.melijn.melijnbot.objects.translation.Translateable
import me.melijn.melijnbot.objects.utils.getDurationString
import me.melijn.melijnbot.objects.utils.getSystemUptime
import me.melijn.melijnbot.objects.utils.sendEmbed
import java.lang.management.ManagementFactory
import java.text.DecimalFormat
import java.util.concurrent.ThreadPoolExecutor


class StatsCommand : AbstractCommand("command.stats") {

    init {
        id = 4
        name = "stats"
        syntax = Translateable("$root.syntax")
        aliases = arrayOf("statistics")
        description = Translateable("$root.description")
        commandCategory = CommandCategory.UTILITY
    }


    override fun execute(context: CommandContext) {
        val bean = ManagementFactory.getPlatformMXBean(OperatingSystemMXBean::class.java)
        val totalMem = bean.totalPhysicalMemorySize shr 20
        val usedMem = totalMem - (bean.freePhysicalMemorySize shr 20)
        val totalJVMMem = ManagementFactory.getMemoryMXBean().heapMemoryUsage.max shr 20
        val usedJVMMem = ManagementFactory.getMemoryMXBean().heapMemoryUsage.used shr 20
        val shardManager = context.getJDA().shardManager
        val voiceChannels = shardManager?.shards?.stream()?.mapToLong { shard ->
            shard.voiceChannels.stream().filter { vc -> vc.members.contains(vc.guild.selfMember) }.count()
        }?.sum()
        val threadPoolExecutor = context.taskManager.getExecutorService() as ThreadPoolExecutor
        val scheduledExecutorService = context.taskManager.getScheduledExecutorService() as ThreadPoolExecutor

        val title1 = Translateable("$root.response.field1.title").string(context)
        val title2 = Translateable("$root.response.field2.title").string(context)
        val title3 = Translateable("$root.response.field3.title").string(context)

        val value1 = replaceValue1Vars(
                Translateable("$root.response.field1.value").string(context),
                shardManager?.shardsTotal ?: 0,
                shardManager?.userCache?.size() ?: 0,
                shardManager?.guildCache?.size() ?: 0,
                voiceChannels ?: 0,
                threadPoolExecutor.activeCount + scheduledExecutorService.activeCount + scheduledExecutorService.queue.size,
                getDurationString(ManagementFactory.getRuntimeMXBean().uptime)
        )

        val value2 = replaceValue2Vars(
                Translateable("$root.response.field2.value").string(context),
                bean.availableProcessors,
                "${usedMem}MB/${totalMem}MB",
                getDurationString(getSystemUptime())
        )

        val value3 = replaceValue3Vars(
                Translateable("$root.response.field3.value").string(context),
                DecimalFormat("###.###%").format(bean.processCpuLoad),
                "${usedJVMMem}MB/${totalJVMMem}MB",
                "${Thread.activeCount()}/${Thread.getAllStackTraces().size}"
        )

        val embed = Embedder(context)
                .setThumbnail(context.getJDA().selfUser.effectiveAvatarUrl)
                .addField(title1, value1, false)
                .addField(title2, value2, false)
                .addField(title3, value3, false)
                .build()

        sendEmbed(context, embed)
    }

    private fun replaceValue1Vars(value: String, shardCount: Int, userCount: Long, guildCount: Long, voiceChannels: Long, threadCount: Int, uptime: String): String {
        return value
                .replace("%shardCount%", shardCount.toString())
                .replace("%userCount%", userCount.toString())
                .replace("%guildCount%", guildCount.toString())
                .replace("%cVCCount%", voiceChannels.toString())
                .replace("%botThreadCount%", threadCount.toString())
                .replace("%botUptime%", uptime)
    }

    private fun replaceValue2Vars(value: String, coreCount: Int, ramUsage: String, uptime: String): String {
        return value
                .replace("%coreCount%", coreCount.toString())
                .replace("%ramUsage%", ramUsage)
                .replace("%systemUptime%", uptime)
    }

    private fun replaceValue3Vars(value: String, cpuUsage: String, ramUsage: String, threadCount: String): String {
        return value
                .replace("%jvmCPUUsage%", cpuUsage)
                .replace("%ramUsage%", ramUsage)
                .replace("%threadCount%", threadCount)
    }
}